pkg/plugins/plugins.go

```diff
@@ -22,6 +29,113 @@ type PluginKey struct {
 	IsInternal     bool
 }
 
+// Implementaion stores implementation type (e.g. instance-storage) and its name (e.g. s3, mongo)
+type Implementaion struct {
+	Type string `json:"type"`
+	Name string `json:"implementation"`
+}
+
+// PluginMetadata about an installed plugin.
+type PluginMetadata struct {
+	Name            string          `json:"name"`
+	ClientPath      string          `json:"clientPath,omitempty"`
+	Implementations []Implementaion `json:"implementations"`
+	VersionInfo
+}
+
+// VersionInfo contains information from running the `version` command against the plugin.
+type VersionInfo struct {
+	Version string `json:"version"`
+	Commit  string `json:"commit"`
+	Author  string `json:"author,omitempty"`
+}
+
+type PluginProvider interface {
+	List() ([]string, error)
+	GetMetadata(string) (*PluginMetadata, error)
+}
+
+func NewFileSystem(config *config.Config) *fileSystem {
```

Comment:
I realize that parts of your PR conflict with mine (#874). Let's merge yours first and I will rationalize the differences in mine.

Reply:
🤦